ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+                masks, scores = self.run_efficient_sam(image, None, box, 1)
+                if len(masks) > 0:
+                    masks_list.append(masks[0])
+                    item_id = f"food_id_{mask_num:d}"
+                    mask_num += 1
+                    mask_msg = self.generate_mask_msg(
+                        item_id, phrase, scores[0], masks[0], image, depth_img, box
+                    )
+                    detected_items.append(mask_msg)
+                    item_labels.append(phrase)
+
+        result.detected_items = detected_items
+        result.item_labels = item_labels
+
+        # Measure the elapsed time running GroundingDINO on the image prompt
+        inference_time = int(round((time.time() - inference_time) * 1000))
+        self._node.get_logger().info(
+            f"Approximate Vision Pipeline Inference Time: {inference_time}"
+        )
+
+        return result
+
+    async def execute_callback(
+        self, goal_handle: ServerGoalHandle
+    ) -> SegmentAllItems.Result:
+        """
+        Execute the action server callback.
+
+        Parameters
+        ----------
+        goal_handle: The goal handle for the action server.
+
+        Returns
+        -------
+        result: The result message containing masks for all food items detected in the image
+                paired with semantic labels.
+        """
+        self._node.get_logger().info("Received a new goal!")
+        starting_time = self._node.get_clock().now()
+
+        # Get the latest image and camera info
+        latest_img_msg = self._node.get_latest_msg(self.rgb_image_topic)
+        if self.camera_info is None:
+            self.camera_info = self._node.get_latest_msg(self.camera_info_topic)
+        if self.camera_info is not None:
+            camera_info = self.camera_info
+        else:
+            camera_info = None
+
+        # Check if the image and camera info are available
+        if latest_img_msg is None or camera_info is None:
+            self._node.get_logger().error("Image or camera info not available.")
+            return SegmentAllItems.Result()
+
+        # Get the caption from the goal request
+        caption = goal_handle.request.caption
+
+        # Create a rate object to control the rate of the vision pipeline
+        rate = self._node.create_rate(self.rate_hz)
+
+        # Define a cleanup function to destroy the rate
+        def cleanup():
+            self._node.destroy_rate(rate)
+
+        # Start running the vision pipeline as a separate thread
+        vision_pipeline_task = self._node.executor.create_task(
+            self.run_vision_pipeline, latest_img_msg, caption
+        )
+
+        # Wait for the vision pipeline to finish and keep publishing
+        # feedback (elapsed time) while waiting
+        feedback = SegmentAllItems.Feedback()
+        while (
+            rclpy.ok()
+            and not goal_handle.is_cancel_requested
+            and not vision_pipeline_task.done()
```

Comment:
Are there cases where the vision pipeline could hang? If so, I'd recommend adding a timeout to the action (maybe in the action message itself) to be robust to that